### PR TITLE
Fixed category filter on Media List and Test MediaAdminController

### PR DESCRIPTION
**Pull request by jordisala1991**
I am targetting this branch, because this is BC.

Closes #1156 #1130 #1054 

## Changelog

```markdown
### Added
- Added test on `MediaAdminController`
### Fixed
- Media List is now filterable by Category again
```

## To do

- [x] Update the tests

## Subject

Same fix as other PR's but with tests.

Not sure about all the mocks I have created. But it will be needed in order to fully test MediaAdminController.

I have done a little bit of refactor on MediaAdminController based on content in CRUDController (changed isGranted by checkAccess, removing some parameters,...)

### CODE REVIEW

File: Controller/MediaAdminController.php [hunk 1]
```diff
@@ -42,22 +41,20 @@ public function createAction(Request $request = null)
     /**
      * {@inheritdoc}
      */
-    public function render($view, array $parameters = array(), Response $response = null, Request $request = null)
+    public function render($view, array $parameters = array(), Response $response = null)
```
Comment (greg0ire) [inline Controller/MediaAdminController.php]: For anyone wondering, this is BC 👍 : https://3v4l.org/LWERG

File: Tests/Controller/MediaAdminControllerTest.php [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+namespace Sonata\MediaBundle\Tests\Controller;
+
+use Prophecy\Argument;
+use Sonata\MediaBundle\Controller\MediaAdminController;
+use Sonata\MediaBundle\Tests\Helpers\PHPUnit_Framework_TestCase;
+
+class MediaAdminControllerTest extends PHPUnit_Framework_TestCase
+{
+    protected $controller;
```
Comment (greg0ire) [inline Tests/Controller/MediaAdminControllerTest.php]: Can and should be private (no big deal though)

File: Controller/MediaAdminController.php (hunk 1 above)
Comment (jordisala1991) [inline Controller/MediaAdminController.php]: It is also overriding a method on CRUD controller (also overriding from Controller), none of the parents have Request as a parameter.

**Comment by jordisala1991** [comment]
Build will fail because of the problem with PHP5.3 and the function call + array access (fixed on #1192).

File: Tests/Controller/MediaAdminControllerTest.php (hunk 2 above)
Comment (jordisala1991) [inline Tests/Controller/MediaAdminControllerTest.php]: will change it

File: Controller/MediaAdminController.php (hunk 1 above)
Comment (core23) [inline Controller/MediaAdminController.php]: This is NOT BC 👎 : https://3v4l.org/dlWbq

Comment (jordisala1991) [inline Controller/MediaAdminController.php]: This is what is actually happening. Im not sure how this works, but CRUDController dont have the last parameter.

The example is the other way around, like @greg0ire said. If someone change my parent class (not sure if thats possible) it is considered BC break? This may be important to another PR I have in mind

Comment (greg0ire) [inline Controller/MediaAdminController.php]: @core23 : I think you are getting it backwards : @jordisala1991 is removing a parameter, which means the child class, if there is any will have one more parameter, not one less.

Comment (core23) [inline Controller/MediaAdminController.php]: Yes, he removes a paramater, but this isn't BC. Adding a parameter with a default value is, but not the other way round.

Comment (greg0ire) [inline Controller/MediaAdminController.php]: It is, see my comment above, with a child class which has one more optional parameter (it's optional because it is after an optional parameter)

Comment (jordisala1991) [inline Controller/MediaAdminController.php]: http://symfony.com/doc/current/contributing/code/bc.html

@core23 @greg0ire are we following Symfony BC promise?

If yes, then this is BC, because removing the last parameter is BC on symfony.

Comment (greg0ire) [inline Controller/MediaAdminController.php]: No we're not, but yes it is BC, as proven in my example. @core23 : your test is for "the other way around"

**Comment by jordisala1991** [comment]
Test done (once Travis pases)

File: Controller/MediaAdminController.php (hunk 1 above)
Comment (greg0ire) [inline Controller/MediaAdminController.php]: @core23 please review again… can we at least agree that you got your example backwards in https://github.com/sonata-project/SonataMediaBundle/pull/1196#discussion_r100724466 ?

Comment (greg0ire) [inline Controller/MediaAdminController.php]: ping @core23 , please don't let this rot. @sonata-project/contributors , if you can chime in…

**Comment by jordisala1991** [comment]
I did a little research on the fourth parameter of render... seems like it was removed because adding it was a BC break.

I am okay with adding again fourth parameter on MedAdminController::render... but then lets define what is BC and what is not, because its not clear and there is a lot of opinion.

https://github.com/sonata-project/SonataAdminBundle/pull/3690

**Comment by greg0ire** [comment]
Can you start a new discussion, preferrably based on a PR you would make on dev-kit?

**Comment by jordisala1991** [comment]
I opened an issue first, will do a PR afterwards.

File: Controller/MediaAdminController.php (hunk 1 above)
Comment (core23) [inline Controller/MediaAdminController.php]: My brain was twisted somehow. Everything fine 👍

**Comment by greg0ire** [comment]
Thanks @jordisala1991 !
